src/snowflake/connector/auth/by_plugin.py

```diff
@@ -89,18 +59,37 @@ class AuthType(Enum):
 class AuthByPlugin(ABC):
     """External Authenticator interface."""
 
-    def __init__(self) -> None:
-        self._retry_ctx = AuthRetryCtx()
+    def __init__(
+        self,
+        timeout: int | None = None,
+        max_retry_attempts: int | None = None,
+        socket_timeout_override: int | None = None,
+        backoff_generator: Iterator | None = None,
+    ) -> None:
         self.consent_cache_id_token = False
-        self._timeout: int = 120
+
+        if timeout is None:
+            timeout = DEFAULT_AUTH_CLASS_TIMEOUT
+
+        self._retry_ctx = TimeoutBackoffCtx(
+            timeout=timeout,
+            max_retry_attempts=max_retry_attempts
+            if max_retry_attempts is not None
+            else int(getenv("MAX_CON_RETRY_ATTEMPTS", DEFAULT_MAX_CON_RETRY_ATTEMPTS)),
+            backoff_generator=backoff_generator,
+        )
+
+        # some authenticators may want to override socket level timeout
+        # for example, AuthByKeyPair will do this to ensure JWT tokens are refreshed in time
+        self._socket_timeout_override = socket_timeout_override
 
     @property
-    def timeout(self) -> int:
-        return self._timeout
+    def timeout(self) -> int | None:
```

Comment:
this is also a BCR, previously the assumption is that `timeout` is always returning an integer, but now it can return None.

code like `AuthByPlugin.timeout + 10` will break now when `AuthByPlugin.timeout` returns None

Reply:
we can guarantee this will not be none